### ISSUE TITLE
[BUG] Inconsitency in qnode results between PennyLane versions.

### ISSUE THREAD

**Issue by cvjjm** [PennyLaneAI/pennylane#2095]
### Expected behavior

The expected behavior is displayed by all PennyLane versions between 0.15 and 0.18 but the asserts fail with 0.19.0, 0.19.1, and 0.20.0 as indicated in the comments. This has been discussed verbally with  @josh146  and @mariaschuld .

### Actual behavior

The behavior of v0.15 to 0.18.

### Source code

```shell
import pennylane as qml
import numpy as np

print(qml.__version__)

dev = qml.device('default.qubit', wires=8)

def ansatz(params):
    # if isinstance(params, qml.numpy.tensor):
    #     params = params.unwrap()
    print(type(params), params)
    qml.RY(2.0943951023931957, wires=[6])
    qml.RY(2.5261129449194057, wires=[4])
    qml.CNOT(wires=[6, 4])
    qml.RY(0.6154797086703873, wires=[4])
    qml.CNOT(wires=[6, 4])
    qml.RY(1.5707963267948966, wires=[0])
    qml.CNOT(wires=[4, 0])
    qml.CNOT(wires=[6, 0])
    qml.RY(-1.5707963267948966, wires=[0])
    qml.CNOT(wires=[4, 0])
    qml.CNOT(wires=[6, 0])
    qml.PauliZ(wires=[0])
    qml.PauliZ(wires=[4])
    qml.PauliZ(wires=[6])
    qml.PauliX(wires=[0])
    qml.CNOT(wires=[0, 5])
    qml.CNOT(wires=[0, 7])
    qml.PauliX(wires=[0])
    qml.PauliX(wires=[4])
    qml.CNOT(wires=[4, 1])
    qml.CNOT(wires=[4, 7])
    qml.PauliX(wires=[4])
    qml.PauliX(wires=[6])
    qml.CNOT(wires=[6, 1])
    qml.CNOT(wires=[6, 5])
    qml.PauliX(wires=[6])
    qml.SWAP(wires=[2, 4]),
    qml.SWAP(wires=[3, 5]),
    qml.CZ(wires=[2, 4]),
    qml.CZ(wires=[3, 5]),
    qml.DoubleExcitation(params[0], wires=[2, 3, 4, 5])
    qml.SWAP(wires=[0, 2]),
    qml.SWAP(wires=[1, 3]),
    qml.CZ(wires=[0, 2]),
    qml.CZ(wires=[1, 3]),
    qml.DoubleExcitation(params[1], wires=[0, 1, 2, 3])
    qml.SWAP(wires=[4, 6]),
    qml.SWAP(wires=[5, 7]),
    qml.CZ(wires=[4, 6]),
    qml.CZ(wires=[5, 7]),
    qml.DoubleExcitation(params[2], wires=[4, 5, 6, 7])
    qml.SWAP(wires=[2, 4]),
    qml.SWAP(wires=[3, 5]),
    qml.CZ(wires=[2, 4]),
    qml.CZ(wires=[3, 5]),
    qml.DoubleExcitation(params[3], wires=[2, 3, 4, 5])

params = np.array([-1.96429246e-02,
                   2.01105944e-05,
                   -7.99301913e-03,
                   2.92601909e-02])


def cost1(params1):
    @qml.qnode(dev, diff_method='parameter-shift')
    def qnode1(loca_params):
        ansatz(loca_params)
        return qml.state()

    ret = qnode1(params1)
    print(qnode1.qtape.operations)
    return ret

def cost2(params2):
    @qml.qnode(dev, diff_method='parameter-shift')
    def qnode2():
        ansatz(params2)
        return qml.state()

    ret = qnode2()
    print(qnode2.qtape.operations)
    return ret

res1 = cost1(params)
res2 = cost2(params)

print(repr(res1))
print(repr(res2))

assert np.allclose(res1, res2), res1 - res2 # passes with 0.15.0 through 0.18.0 and with >= 0.20.0 but fails with 0.19.0 and 0.19.1 (res2 remains correct, but res1 is wrong)


def cost3(params1):
    @qml.qnode(dev, diff_method='parameter-shift')
    def qnode3(loca_params):
        ansatz(loca_params)
        return [qml.expval(qml.PauliX(0)), qml.expval(qml.PauliX(1)), qml.expval(qml.PauliX(0) @ qml.PauliX(1) @ qml.PauliX(2) @ qml.PauliX(3))]

    ret = qnode3(params1)
    print(qnode3.qtape.operations)
    return ret

res3 = cost3(params)

print(repr(res3))
assert np.allclose(res3, [0., 0.,  4.89587945e-01]) # passes with 0.15.0 through 0.18.0 but fails with 0.19.0, 0.19.1, and 0.20.0 with both diff_method='parameter-shift' and without, it works with qml.qnode_old.qnode()


# check with of res1 or res2 are wrong
expected_res12 = qml.numpy.tensor([8.32667268e-17+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        6.80649650e-03+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        7.04504103e-03+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        4.99969272e-01+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 1.11022302e-16+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 8.32667268e-17+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        2.82588467e-03+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        4.89578216e-01+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        7.14310823e-01+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j, 0.00000000e+00+0.j, 0.00000000e+00+0.j,
        0.00000000e+00+0.j], requires_grad=True)

assert np.allclose(res2, expected_res12)
assert np.allclose(res1, expected_res12) # passes with 0.15.0 through 0.18.0 and with >= 0.20.0 but fails with 0.19.0 and 0.19.1 (res2 remains correct, but res1 is wrong)
```

### System information

```shell
This has been tested with different PL versions so this would be of little use.
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

**Comment by josh146** [comment]
Thanks @cvjjm! I've tracked this down to an errant `if` condition, that has been fixed in #2097

**Comment by cvjjm** [comment]
I can confirm that with #2097 the above example works again!
